RecoCTPPS/ProtonReconstruction/plugins/CTPPSProtonProducer.cc

```diff
@@ -0,0 +1,210 @@
+  multiRPReconstructionLabel_ (iConfig.getParameter<std::string>("multiRPReconstructionLabel")),
+  algorithm_                  (iConfig.getParameter<bool>("fitVtxY"), iConfig.getParameter<bool>("useImprovedInitialEstimate"), verbosity_),
+  opticsValid_(false),
+  currentCrossingAngle_(-1.)
+{
+  if (doSingleRPReconstruction_)
+    produces<reco::ForwardProtonCollection>(singleRPReconstructionLabel_);
+
+  if (doMultiRPReconstruction_)
+    produces<reco::ForwardProtonCollection>(multiRPReconstructionLabel_);
+}
+
+//----------------------------------------------------------------------------------------------------
+
+void CTPPSProtonProducer::fillDescriptions(edm::ConfigurationDescriptions& descriptions)
+{
+  edm::ParameterSetDescription desc;
+  desc.addUntracked<unsigned int>("verbosity", 0)->setComment("verbosity level");
+  desc.add<edm::InputTag>("tagLocalTrackLite", edm::InputTag("ctppsLocalTrackLiteProducer"))
+    ->setComment("specification of the input lite-track collection");
+
+  desc.add<bool>("doSingleRPReconstruction", true)
+    ->setComment("flag whether to apply single-RP reconstruction strategy");
+  desc.add<std::string>("singleRPReconstructionLabel", "singleRP")
+    ->setComment("output label for single-RP reconstruction products");
+
+  desc.add<bool>("doMultiRPReconstruction", true)
+    ->setComment("flag whether to apply multi-RP reconstruction strategy");
+  desc.add<std::string>("multiRPReconstructionLabel", "multiRP")
+    ->setComment("output label for multi-RP reconstruction products");
+
+  desc.add<bool>("fitVtxY", true)
+    ->setComment("for multi-RP reconstruction, flag whether y* should be free fit parameter");
+  desc.add<bool>("useImprovedInitialEstimate", true)
+    ->setComment("for multi-RP reconstruction, flag whether a quadratic estimate of the initial point should be used");
+
+  descriptions.add("ctppsProtons", desc);
+}
+
+//----------------------------------------------------------------------------------------------------
+
+void CTPPSProtonProducer::produce(edm::Event& iEvent, const edm::EventSetup& iSetup)
+{
+  // get conditions
+  edm::ESHandle<LHCInfo> hLHCInfo;
+  iSetup.get<LHCInfoRcd>().get(hLHCInfo);
+
+  edm::ESHandle<LHCInterpolatedOpticalFunctionsSetCollection> hOpticalFunctions;
+  iSetup.get<CTPPSInterpolatedOpticsRcd>().get(hOpticalFunctions);
+
+  // re-initialise algorithm upon crossing-angle change
+  if (hLHCInfo->crossingAngle() != currentCrossingAngle_) {
+    currentCrossingAngle_ = hLHCInfo->crossingAngle();
+
+    if (hOpticalFunctions->empty()) {
+      edm::LogWarning("CTPPSProtonProducer") << "No optical functions available, reconstruction disabled.";
+      algorithm_.release();
+      opticsValid_ = false;
+    }
+    else {
+      algorithm_.init(*hOpticalFunctions);
+      opticsValid_ = true;
+    }
+  }
+
+  // stop if conditions invalid
+  if (!opticsValid_)
+    return;
```

Comment:
the producer should still create an (empty) output.

Reply:
Done.